s3prl/downstream/speech_translation/expert.py

```diff
@@ -0,0 +1,506 @@
+                eg. downstream/example/config.yaml
+
+            expdir: string
+                The expdir from command-line argument, you should save all results into
+                this directory, like some logging files.
+
+            **kwargs: dict
+                All the arguments specified by the argparser in run_downstream.py
+                and all the other fields in config.yaml, in case you need it.
+                
+                Note1. Feel free to add new argument for __init__ as long as it is
+                a command-line argument or a config field. You can check the constructor
+                code in downstream/runner.py
+        """
+
+        super(DownstreamExpert, self).__init__()
+
+        print(downstream_expert)
+
+        self.src_lang = downstream_expert['src_lang']
+        self.tgt_lang = downstream_expert['tgt_lang']
+        self.post_process = downstream_expert['post_process']
+        self.output_prefix = downstream_expert['output_prefix']
+        self.upstream_rate = upstream_rate
+
+        self.datarc = downstream_expert['datarc']
+        self.max_positions = downstream_expert['modelrc']['max_source_positions']
+
+
+        self.task = S3prl_SpeechToTextTask.setup_task(Namespace(**downstream_expert['taskrc']))
+        self.task.upstream_rate = upstream_rate
+
+        self.data_dir = downstream_expert['taskrc']['data']
+
+        self.criterion = self.task.build_criterion(Namespace(**downstream_expert['criterionrc']))
+
+        modelrc = Namespace(**downstream_expert['modelrc'])
+        assert modelrc.arch in fairseq.models.ARCH_CONFIG_REGISTRY
+        fairseq.models.ARCH_CONFIG_REGISTRY[modelrc.arch](modelrc)
+        self.model = self.task.build_model(modelrc, upstream_dim)
+        
+        print(self.model)
+
+        self.generator = self.task.build_generator([self.model], Namespace(**downstream_expert['generatorrc']))
+        self.batch_itr = {}
+        
+        self.use_asr = downstream_expert['taskrc']['use_asr']
+
+        if self.use_asr:
+
+            rc = downstream_expert['asrrc']
+            self.asr_datarc = rc['datarc']
+            self.asr_weight = rc['weight']
+            self.asr_dict = Dictionary.load(f"{self.data_dir}/{rc['vocab_file']}")
+            asr_bperc = rc['bpe_tokenizer'].copy()
+            asr_bperc['sentencepiece_model'] = f"{self.data_dir}/{asr_bperc['sentencepiece_model']}" 
+            self.asr_bpe = encoders.build_bpe(Namespace(**asr_bperc))
+            
+            self.asr_decoder = TransformerDecoderScriptable(
+                self.model.decoder.args,
+                self.asr_dict,
+                Embedding(len(self.asr_dict), self.model.decoder.embed_dim, self.asr_dict.pad())
+            )
+            self.asr_task = S3prl_SpeechToTextTask.setup_task(Namespace(**downstream_expert['taskrc']))
+            self.asr_task.tgt_dict = self.asr_dict
+            self.asr_model = S2TTransformerModel(self.model.encoder, self.asr_decoder)
```

Comment:
It's good to direct add an additional decoder for ASR. Another option would be to directly add a CTC layer to the encoder so that we won't be burdened too much for a large decoder. In espnet, we mostly do with the second way for efficient purpose.

Reply:
I think adding a CTC layer is a better way too and I will change the setting.